File: Iteration 1/code/CSE3063GRP12/src/features/Advisor/CourseApprovalController.java
package features.advisor;

import core.general_providers.TerminalManager;
import core.repositories.CourseEnrollmentRepository;
import core.models.concretes.CourseEnrollment;
import core.exceptions.UnexpectedInputException;
import features.approval_courses_selected.ApprovalCoursesSelectedController;
import features.main_menu.MenuController;
import java.util.ArrayList;

public class CourseApprovalController {
    private CourseApprovalView courseApprovalView;
    private CourseEnrollmentRepository courseEnrollmentRepository;

    public CourseApprovalController() {
        courseEnrollmentRepository = new CourseEnrollmentRepository();
        courseApprovalView = new CourseApprovalView();

        try {
            handleApprovalController();
        } catch (UnexpectedInputException e) {
            courseApprovalView.showErrorMessage(e);
        }
    }

    private ArrayList<CourseEnrollment> fetchPendingEnrollments() {
        return courseEnrollmentRepository.getPendingEnrollments();

    }

    private void navigateToApprovalCoursesSelected(CourseEnrollment courseEnrollment) {

        new ApprovalCoursesSelectedController(courseEnrollment);
    }

    private void navigateToMenu() {
        new MenuController();
    }

    private String getUserInput() {
        String input = TerminalManager.getInstance().read();
        // TerminalManager.getInstance().dispose();
        return input;
    }

    private void handleApprovalController() throws UnexpectedInputException {
        ArrayList<CourseEnrollment> pendingEnrollments = fetchPendingEnrollments();
        courseApprovalView.showPendingCourseEnrollments(pendingEnrollments);

        courseApprovalView.showPromptMessage();
        String selection = getUserInput();
        if (selection.matches("^\\d+$") == false && selection.equals("q") == false) {
            throw new UnexpectedInputException();
        } else if (selection.equals("q")) {
            navigateToMenu();
        } else {
            int index = Integer.parseInt(selection);
            if (index > courseEnrollmentRepository.getPendingEnrollments().size() || index < 0) {
                throw new UnexpectedInputException();
            } else {
                navigateToApprovalCoursesSelected(courseEnrollmentRepository.getPendingEnrollments().get(index - 1));
            }
        }
    }
}

File: Iteration3/code/CSE3063GRP12/src/features/advisor/course_approval/CourseApprovalController.py
from core.repositories.CourseEnrollmentRepository import CourseEnrollmentRepository
from core.exceptions.UnexpectedInputException import UnexpectedInputException
from features.advisor.approval_courses_selected.ApprovalCoursesSelectedController import ApprovalCoursesSelectedController
from features.advisor.course_approval.CourseApprovalView import CourseApprovalView


class CourseApprovalController:
    def __init__(self):
        self.__courseEnrollmentRepository = CourseEnrollmentRepository()
        self.__courseApprovalView = CourseApprovalView()

        try:
            self.handle_approval_controller()
        except UnexpectedInputException as e:
            self.__courseApprovalView.showErrorMessage(e)
            CourseApprovalController()

    def fetch_pending_enrollments(self):
        return self.__courseEnrollmentRepository.get_pending_enrollments()

    def navigate_to_approval_courses_selected(self, course_enrollment):
        ApprovalCoursesSelectedController(course_enrollment)

    def navigate_to_menu(self):
        from features.main_menu.MenuController import MenuController
        MenuController()

    def handle_approval_controller(self):
        pending_enrollments = self.fetch_pending_enrollments()
        self.__courseApprovalView.showPendingCourseEnrollments(pending_enrollments)


        selection = input("Please enter your selection: ")
        if not selection.isdigit() and selection != "q":
            raise UnexpectedInputException()
        elif selection == "q":
            self.navigate_to_menu()
        else:
            index = int(selection)
            if index > len(self.__courseEnrollmentRepository.get_pending_enrollments()) or index < 0:
                raise UnexpectedInputException()
            else:
                self.navigate_to_approval_courses_selected(self.__courseEnrollmentRepository.get_pending_enrollments()[index - 1])
